File: manager-app/app/src/main/java/io/github/chenyurumeng/aghmanager/RootShell.java
package io.github.chenyurumeng.aghmanager;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;

final class RootShell {
    static final class Result {
        final int code;
        final String output;

        Result(int code, String output) {
            this.code = code;
            this.output = output == null ? "" : output.trim();
        }

        boolean ok() {
            return code == 0;
        }
    }

    static Result exec(String command) {
        return exec(command, 30);
    }

    static Result exec(String command, long timeoutSeconds) {
        Process process = null;
        StringBuilder out = new StringBuilder();
        try {
            process = new ProcessBuilder("su", "-c", command)
                    .redirectErrorStream(true)
                    .start();

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (out.length() > 0) out.append('\n');
                    out.append(line);
                }
            }

            if (!process.waitFor(timeoutSeconds, TimeUnit.SECONDS)) {
                process.destroyForcibly();
                return new Result(124, "Command timed out after " + timeoutSeconds + "s");
            }
            return new Result(process.exitValue(), out.toString());
        } catch (Exception e) {
            return new Result(127, e.getClass().getSimpleName() + ": " + e.getMessage());
        } finally {
            if (process != null) process.destroy();
        }
    }

    private RootShell() {}
}

File: manager-app/app/src/main/java/io/github/chenyurumeng/aghmanager/data/RootShell.kt
package io.github.chenyurumeng.aghmanager.data

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit

data class ShellResult(
    val exitCode: Int,
    val stdout: String,
    val timedOut: Boolean = false
) {
    val ok: Boolean
        get() = exitCode == 0 && !timedOut
}

object RootShell {
    suspend fun exec(command: String, timeoutSeconds: Long = 30): ShellResult =
        withContext(Dispatchers.IO) {
            var process: Process? = null
            try {
                process = ProcessBuilder("su", "-c", command)
                    .redirectErrorStream(true)
                    .start()

                coroutineScope {
                    val reader = async(Dispatchers.IO) {
                        process.inputStream.bufferedReader(StandardCharsets.UTF_8).use { it.readText() }
                    }

                    val finished = process.waitFor(timeoutSeconds, TimeUnit.SECONDS)
                    if (!finished) {
                        process.destroyForcibly()
                        val partial = withTimeoutOrNull(2_000L) { reader.await() }.orEmpty()
                        ShellResult(
                            exitCode = 124,
                            stdout = if (partial.isBlank()) {
                                "Command timed out after " + timeoutSeconds + "s"
                            } else {
                                partial.trim()
                            },
                            timedOut = true
                        )
                    } else {
                        ShellResult(
                            exitCode = process.exitValue(),
                            stdout = reader.await().trim()
                        )
                    }
                }
            } catch (cancelled: CancellationException) {
                process?.destroyForcibly()
                throw cancelled
            } catch (e: Exception) {
                ShellResult(
                    exitCode = 127,
                    stdout = e.javaClass.simpleName + ": " + (e.message ?: "")
                )
            } finally {
                process?.destroy()
            }
        }
}
